src/emqtt.erl

```diff
@@ -828,15 +843,30 @@ mqtt_connect(State = #state{clientid    = ClientId,
                                  username     = Username,
                                  password     = emqtt_secret:unwrap(Password)}), State).
 
-reconnect(state_timeout, NextTimeout, #state{conn_mod = CMod} = State) ->
-    case do_connect(CMod, State#state{clean_start = false}) of
```

Comment:
Previously, reconnecting would set clean_start to false. I don't quite understand the reason for this, it doesn't seem to make sense 🤔.

So I removed this logic and resent inflight messages when `clean_start = false` only, otherwise discard these messages.

Reply:
The reconnect function was added for client to reconnect and resume the session on the broker when the transport connection is unexpectedly closed. 

That is both client and broker have a state that must be kept after reconnecting, aka transport layer recovery.
Client could receive the messages when it is offline and the Server get buffered messages while transport was down.

My understanding is at the client side,  reconnect with clean_start = true  has same effects as starting one new client process. So Client wants a clean session after the transport disconnect should not enable reconnect option.

Reply:
Is there a way to let client connect with clean start = True then reconnect with clean start = False in case of disconnect?

Reply:
> Is there a way to let client connect with clean start = True then reconnect with clean start = False in case of disconnect?

In `clean_start = false`, the subscription relationship is stored on the broker side (in emqx, it is kept for 2 hours by default). This will cause the caller to get the last connection subscription relationship or messages if they restart the emqtt process with `clean_start = false`. **Therefore, we should not modify the value of clean_start, which may cause unpredictable situations for the caller.**

Reply:
> My understanding is at the client side, reconnect with clean_start = true has same effects as starting one new client process. So Client wants a clean session after the transport disconnect should not enable reconnect option.

Can't agree any more.  With `clean_start = true`, only restoring the connection is not enough now? The subscription relationships established by the caller are also lost after reconnection unless we re-initiate those subscriptions again 🙃.

Can we **(A** disable the feature that starts emqtt with the `clean_start=true` and `reconnect=true` parameters, and suggest the caller to start with `clean_start=false`?

Or still **(B** reconnect with the `clean_start=true` value. But resend the inflight messages, even automatically re-initiate the subscription request (just working like `clean_start=false`). The only problem is that the PUBLISH or PUBREL messages in inflight are retransmitted. This may cause some duplicate messages for the broker, or unknown PUBREL messages.

I prefer **(A** because it is simple and straightforward enough :)